content/blog/eng/automation-testing-a-step-back.md

```diff
@@ -0,0 +1,114 @@
+
+This may seem like a marginal problem but having a slow CI is problematic. So be careful because at the beginning they can be like grains of sand and when the project becomes very large, the CI can become like a sandstorm and block the development process like a slow build.
+
+## Make harder maintenance and refactoring
+
+This is the case when a change to the codebase (like refactoring an implementation detail or adding a new feature) also implies changes to the test code even if the previous tested behavior doesn't change.
+
+This means that your test code is heavily coupled with your tested code. And can make frustrating, slow and inefficient working on a project.
+
+## Make write worst code
+
+This point is the exact opposite of the one written above on the list of the pros. Indeed to make a portion of code testable we are often forced to restructure our code, but we can harm doing that.
+
+This is called  _"test-induced design damage"_. This is a concept well expressed [here](https://dhh.dk/2014/test-induced-design-damage.html), that explains the process when the pursuit of easier, faster, or more testing warps your code through needless indirection or complexity.
+
+# Many ways of testing (wrong)
+
+Doing automation test is pretty complex. The first time I entered the testing world it was a nightmare, I was overwhelmed by how many concepts are needed only for testing purposes (things like all types of tests, test-driven development (TDD), test doubles, etc). And on top of that all the framework related stuff. So that I asked myself: is it worth it? Short answer: if you want to be a good developer, yes.
+
+So there are many theoretical concepts and many ways of applying them, and it's really very simple to do it the wrong way.
+
+For pure coding related concepts it's more simple to understand what we are doing wrong. Leaving out the functional aspect, because it is obvious that the code that works as we expect is better than the code that does not. Let's analyze it from an architectural point of view.
+
+Like I said before the code that is easy to change even after months is better than the code that is harder to change, the code that doesn't break in point A when a point B is changed is better than the one how does, etc. So we can put ourselves in a position when, based on experience, we can define boundaries to respect our standard. And should be pretty clear when we cross those boundaries.
+
+For test is a bit different, typically the testing code is not treated as source code, but _"__Test code  **is**  source code, you need to architect it, to maintain it, to refactor it. It's source code that you are not shipping to your users"._ Christina Lee
+
+# How should we test?
+
+Above you can find the reason why you should test and why you should do it carefully. It is really important to do it right or it could be more bad than good. But what are the 'rules' to do it right?
```

Comment:
It is really important to do it right, otherwise it can be more harmful than useful.